machines: Boot order, cpu type and emulated machine

mareklibra opened the pull request:
The VM Overview tab newly renders:

- CPU Type (like 'Conroe' or 'Haswell')
- Emulated Machine (like 'pc-i440fx-rhel7.3.0')
- Boot Order - (like 'cdrom,disk,network')
    
Following properties are removed from rendering:

- VM ID
- OS Type (Libvirt's support is limitted, so useless)
- Status (duplicated with listing row status icon)

In addition to this change, an external provider can enrich

- the action list for provider-specific buttons (along recent 'Shut down' or 'Run')
- VM properties displayed on the Overview subtab

![reviewedvmprops](https://cloud.githubusercontent.com/assets/17194943/24204856/05f5db12-0f1a-11e7-97b6-f787036978a1.png)

andreasn commented:
Looks good in general.

One small thing that looks weird to me is that "State" doesn't align to the top, so that makes "Emaluted" machine appear much higher up.
![screenshot from 2017-03-23 17-11-30](https://cloud.githubusercontent.com/assets/209895/24257838/35666d4e-0fec-11e7-8712-0e0ae525fd08.png)

mareklibra commented:
@andreasn , vertical alignment is fixed.
![vmoverview_fixedverticalalignment](https://cloud.githubusercontent.com/assets/17194943/24284788/554e1874-106d-11e7-8c8f-0777ead2da77.png)

andreasn commented:
Thank you for fixing that so quickly!
Looks good to me!

mareklibra commented:
Failing test on debian-8 seems to be environment-related and should pass after restart.
Remaining failing tests are unrelated.

mykaul commented:
What do 'autostart' and 'boot order' have to do with 'Emulated machine' and 'CPU type' ? Why are they grouped together?

mareklibra commented:
@mykaul , do you mean within single pull request? This PR is result of general displayed properties review.
The focus is too limit their count to the most important only. Any suggestion what to add/remove, please?

Considering recent set of rendered properties, I chose two columns on the Overview tab to save space for provider's data (like from oVirt).

mykaul commented:
@mareklibra - I'm not talking about the PR but about the UI look and feel. The grouping makes no logical sense. Right now, you have enough space on the right - and if you plan to include more items, please design the dialog first of what you'd like to present.
(State: 'running' is redundant, btw).

What doe 'Connection: system' mean? What other options are there?

mareklibra commented:
I would like to reserve the third column for provider (like oVirt) specific content which is good-to-have on the Overview subtab (like description, comments or icon, see older [screenshot for oVirt](https://trello-attachments.s3.amazonaws.com/57d928d530d7cc8eeb4af00b/588f0dff2b3e7dbc9a0a3ab1/7adec0efdacfbc2aed18a96426dbc36e/cockpitMachinesExtProvider_hostVmsOverview.png)).

I can move the Emulated machine and CPU Type to separate column but it will increase requirement on screen width.
Unless even more properties are identified for rendering, I would not add anything to the first column, since vm-related error messages are recently rendered here (sort of "Run VM failed", etc.).

Recently there's no plan to add more properties to the Overview subtab.

Regarding the "Connection": it's libvirt's connection name (system/session). Can be enriched for provider's specifics, like 'oVirt' or other source name from user perspective.

mareklibra commented:
@mykaul , @garrett , so the proposed change:
- just 2 columns are still used for Libvirt
- 3rd column is reserved for optional provider's extension
- the duplicated "Status" field is removed
- Emulated Machine and CPU Type are moved to the first column
- potential VM-related error message is rendered under the columns (spanning whole row)

Please let me know, if it makes sense.

dperpeet commented on pkg/machines/helpers.es6:
```diff
@@ -108,7 +108,18 @@ const transform = {
+        cdrom: _("cdrom"),
+        'interface': _("network"),
+        hd: _("disk")
+
```
extra whitespace

mareklibra replied:
done

dperpeet commented on pkg/machines/helpers.es6:
```diff
@@ -108,7 +108,18 @@ const transform = {
+    'bootableDisk': {
+        disk: _("disk"),
+        cdrom: _("cdrom"),
+        'interface': _("network"),
```
why is only `interface` here in quotes?

mareklibra replied:
The others are not needed, but the 'interface' is reserved word.
For consistency, I'm adding quotes for the remaining as well.

dperpeet commented on pkg/machines/hostvmslist.jsx:
```diff
@@ -38,47 +38,64 @@ const NoVm = () => {
+    let providerActions = null;
+    if (config.provider.vmActionsFactory) {
+        const ProviderActions = config.provider.vmActionsFactory();
+        providerActions = (<ProviderActions vm={vm} providerState={config.providerState} dispatch={dispatch} />);
```
since this is a single line, I think I would skip the braces `(`, `)`

mareklibra replied:
done

dperpeet commented on pkg/machines/libvirt.es6:
```diff
@@ -343,6 +348,80 @@ function parseDumpxmlForDisks(devicesElem) {
+            const bootElem = getSingleOptionalElem(deviceElem, 'boot');
+            if (bootElem && bootElem.getAttribute('order')) {
+                bootableDevices.push({
+                    // TODO: add additional device identification/details if needed, so far just the 'type' is rendered
```
Does this need to be a `TODO`? Is something missing?

mareklibra replied:
ok, rephrasing and removing TODO

dperpeet commented on test/verify/check-machines:
```diff
@@ -202,6 +204,9 @@ class TestMachines(MachineCase):
 
+        b.wait_present("#test-vm-action-{0}".format(self.providerVm1_name)) # is provider-specific action button rendered?
+        b.wait_in_text("#test-vm-props-{0}".format(self.providerVm1_name), "Test Provider Property") # are provider-specific VM props rendered?
+
```
Is it possible to add testing for the cpu type and emulated machine?

mareklibra replied:
Sure, I'll see how far I can get

mareklibra replied:
@dperpeet , please have a look.
Unless reading the cpu type/emulated machine from the machine and comparing it back to UI (which is probably not needed), I'm unable to check for appropriate value.
I believe this is still sufficient:
- cpu type is either "host" (as expected from virt-install params) or "custom" (for debian-8)
- emulated machine value is asserted to be non-empty (depends on test machine)

dperpeet replied:
this is sufficient - at least we're checking that there is info in those fields

mareklibra commented:
PR updated for the review comments above and rebased.

UI has changed, please have a look at the recent screenshot.
I removed the duplicated "Status" and restructured the properties into 3(4) columns.
The 3(4) columns require wider screen now but look better than unbalanced in just 2(3) as proposed above.
![vmoverview](https://cloud.githubusercontent.com/assets/17194943/24749658/3593202c-1ac4-11e7-8e8f-0e4c76dfc7fb.png)

mykaul commented:
- What is 'Connection' and why is it 'System'? Why is it even important for
the user?
- I'm not sure 'shut off' is the right state. 'Down' is the usual term?
- 'vCPU' not left-aligned (not aligned with memory above it)? Same goes to
'CPU Type' and 'Autostart'
- If there is a description field for disks, I think it's more important
than many of the fields.
- Is the an icon for the VM?
- OS Type?
- What is 'Usage' ?
- Is the title 'Virtual Machines' needed? A bit redundant with the menu on
the left and takes screen estate.
- How does it look with a list of (for example) 10 VMs?



On Thu, Apr 6, 2017 at 1:27 PM, Marek Libra <notifications@github.com>
wrote:

> PR updated for the review comments above and rebased.
>
> UI has changed, please have a look at the recent screenshot.
> I removed the duplicated "Status" and restructured the properties into
> 3(4) columns.
> The 3(4) columns require wider screen now but look better than unbalanced
> in just 2(3) as proposed above.
> [image: vmoverview]
> <https://cloud.githubusercontent.com/assets/17194943/24749658/3593202c-1ac4-11e7-8e8f-0e4c76dfc7fb.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockpit-project/cockpit/pull/6181#issuecomment-292132990>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEcJ6ZoqCi0aDsRIimpYIo1EYOBvnWi2ks5rtL4QgaJpZM4MlWav>
> .
>

mareklibra commented:
@mykaul , this is Libvirt-based version without oVirt data enrichment.

With this:
- Connection is either System or Session, they are intentionally mixed in one list, sorting is planed. User needs to use it at least as a namespace.
- The 'shut off' follows Libvirt's terminology, not Ovirt
- the property names are right-aligned reusing Patternfly CSS. @andreasn , does it conform the rest of Cockpit?
- The screenshot shows potential vm-related error message as well - how/where is it rendered. In such case, the vm name (listing row) is enriched for the warning icon to differentiate it from others even in collapsed state.
- OS Type is hard to get in Libvirt implementation, so far there's no communication with agent.

With optional cockpit-machines-ovirt-provider installed, the vm description, icon, templates, OS Type and others are displayed here or on additional subtabs as well. Plus ovirt-specific actions are allowed.

mareklibra commented:
![vmlist](https://cloud.githubusercontent.com/assets/17194943/24751176/df02d2c0-1ac8-11e7-9c09-8013b767c4d1.png)

![vmlistexpanded](https://cloud.githubusercontent.com/assets/17194943/24751181/e261df24-1ac8-11e7-881c-7476d2d391d2.png)

dperpeet commented:
This looks good now. @garrett any objections?

garrett commented:
1. it's odd to have the error status in the VM's name — shouldn't prefixing the state with the error be more appropriate?
2. I would consider it more of an error than a warning (a warning icon is currently used)
3. @andreasn and I have some concerns with the inline error message that looks like it's just "hanging out" oddly

(I'm working on a mockup to address these concerns; wanted to get them to you in textual form ASAP though.)

garrett commented:
Also:

4.  "**VM START action failed**" sounds like an error code, not an error message.

mykaul commented:
I don't see why Connection is such an important property. Can you explain?

mareklibra commented:
It would be not possible to differentiate between two VMs of the same name without knowledge of connection.
The user might not have permissions to issue active actions on System VMs but should be able to see them.
Column sorting is planed - user will be able the simply split the list if needed.

VMs running on the host can originate from multiple sources.
The connection will state "oVirt" if the VM is managed by oVirt to clearly differentiate the "source" (not yet implemented). Other providers might add their own values.

garrett commented:
![action-fail-3](https://cloud.githubusercontent.com/assets/10246/24761763/a77d8632-1aec-11e7-8a3b-ac65cbbece2b.png)

This mockup ensures that the starting VM error is shown even when the row isn't expanded by introducing an error state. It also shows a message with more detail (although I think the text should still be worked on in the notification bar).

What does everyone think? (Especially @mareklibra & @andreasn)

mareklibra commented:
@garrett , I like it, looks better.

How can the user check actual VM status in case of an error message is shown?
I just removed the Status from the Overview subtab.

mykaul commented:
@mareklibra - I still don't see the connection type as the 2nd most important item to display. I also don't think the user would actually understand the differences. If anything, something like '2vCPUs/2GB' RAM would be much more relevant, especially when the list is collapsed. Or split vCPUs count and memory and display both in different coliuns.

garrett commented:
@mareklibra: What do you mean by "How can the user check actual VM status in case of an error message is shown?"

Could there be any other status besides stopped (& error) if there are problems starting the VM?

mareklibra commented:
@garrett , Yes, the VM can be in paused state. Not sure, but theoretically even in "running", no matter the action failed.

Same pattern will be used for other VM actions as well where complexity of potential UI states grows.
The user shall be able to see libvirt's VM state.

dperpeet commented:
I think the open design issues can be addressed in a follow-up and aren't strictly related to this pr.
